### PR TITLE
Stabilize the copy_closures and clone_closures features

### PR DESCRIPTION
**Pull request by SimonSapin**
In addition to the `Fn*` family of traits, closures now implement `Copy` (and similarly `Clone`) if all of the captures do.

Tracking issue: https://github.com/rust-lang/rust/issues/44490

### CODE REVIEW

File: src/librustc/traits/select.rs [hunk 1]
```diff
@@ -2086,12 +2086,8 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
...
             ty::TyClosure(def_id, substs) => {
                 let trait_id = obligation.predicate.def_id();
-                let copy_closures =
-                    Some(trait_id) == self.tcx().lang_items().copy_trait() &&
-                    self.tcx().has_copy_closures(def_id.krate);
-                let clone_closures =
-                    Some(trait_id) == self.tcx().lang_items().clone_trait() &&
-                    self.tcx().has_clone_closures(def_id.krate);
+                let copy_closures = Some(trait_id) == self.tcx().lang_items().copy_trait();
+                let clone_closures = Some(trait_id) == self.tcx().lang_items().clone_trait();
```
Comment (SimonSapin) [inline src/librustc/traits/select.rs]: I’m not confident about this particular chunk, but tests appear to be happy.

File: src/libcore/marker.rs [hunk 2]
```diff
@@ -166,6 +166,11 @@ pub trait Unsize<T: ?Sized> {
 /// are allowed to access `x` after the assignment. Under the hood, both a copy and a move
 /// can result in bits being copied in memory, although this is sometimes optimized away.
 ///
+/// ## Closures
+///
+/// Closure types automatically implement `Clone` if they capture no value from the environment
```
Comment (chpio) [inline src/libcore/marker.rs]: > `Clone`

is that a typo for `Copy`?

Comment (SimonSapin) [reply src/libcore/marker.rs]: Oops, hasty copy-paste. Good catch, thanks. I just squashed a fix.

File: src/librustc/traits/select.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc/traits/select.rs]: seems ok. The name of the variable is weird; I would prefer `is_copy_trait` and `is_clone_trait` or something.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
do you mind changing the names of those variables? otherwise r=me :)

**Comment by SimonSapin** [comment]
Yes, this code is much less confusing with those variable names. Done, thanks.

**Pull request merged**